src/main/jbake/content/blog/2017/09/agent-to-services-over-ssl.adoc

```diff
@@ -0,0 +1,146 @@
+= Connecting Hawkular Agent to Hawkular Services over SSL
+Josejulio Martínez
+2017-09-11
+:jbake-type: post
+:jbake-status: published
+:jbake-tags: blog, services, ssl
+
+SSL provides identification of the communicating parties, confidentiality and integrity of the information being shared.
+
+In production environments, network eavesdropping could be a concern. You can use SSL to provide a secure communication
+channel between servers.
+
+Hawkular Services expects Hawkular Agents to connect, push information and listen for commands.
+These commands also include server credentials that could be vulnerable to a man-in-the-middle attack.
+
+A previous link:http://www.hawkular.org/hawkular-services/docs/installation-guide/secure-comm.html[article] shows how
+to configure your Hawkular Services with SSL and have Hawkular Agents connect to it trough SSL.
+This guide will show how is done for Dockerized Hawkular Services Agents.
+
+== Preparing your certificates
+
+Before starting we need to prepare the certificates that we are going to use.
+Your public and private key for Hawkular Services need to be on PEM or PKC12 format.
+For this guide we will use PEM.
+
+We can create self-signed certificates (in PEM format) using:
+```
+keytool -genkey -keystore hawkular.keystore -alias hawkular -dname "CN=hawkular-services" -keyalg RSA -storepass hawkular -keypass hawkular -validity 36500 -ext san=ip:127.0.0.1,dns:hawkular-services
+keytool -importkeystore -srckeystore hawkular.keystore  -destkeystore hawkular.p12 -deststoretype PKCS12 -srcalias hawkular -deststorepass hawkular -destkeypass hawkular -srcstorepass hawkular
+openssl pkcs12 -in hawkular.p12 -password pass:hawkular -nokeys -out hawkular-services-public.pem
+openssl pkcs12 -in hawkular.p12 -password pass:hawkular -nodes -nocerts -out hawkular-services-private.key
+```
+
+IMPORTANT: When creating the certificates, remember to include the host that Hawkular will use
+(as Common Name(CN) and Subject Alternative Name(SAN)), else the certificate validation will fail.
+Through this guide, host will be set to hawkular-services.
+
+WARNING: Always include san=ip:127.0.0.1 in your certificate, as this will be used internally by Hawkular Services.
+
+== Starting Cassandra
+
+Hawkular Services requires a cassandra instance, you can start one by doing:
+
+```
+docker run --name hawkular-cassandra -e CASSANDRA_START_RPC=true -d cassandra:3.0.9
+```
+
+== Starting Hawkular Services on SSL
+
+Now that there is a cassandra ready, Hawkular Services can be started,
+it will be linked to the cassandra instance (hawkular-cassandra)
+
+```
+docker run --name hawkular-services --link=hawkular-cassandra -e CASSANDRA_NODES=hawkular-cassandra -e HAWKULAR_HOSTNAME=hawkular-services -e HAWKULAR_USE_SSL=true -p 8443:8443 -v `pwd`/hawkular-services-private.key:/client-secrets/hawkular-services-private.key -v `pwd`/hawkular-services-public.pem:/client-secrets/hawkular-services-public.pem hawkular/hawkular-services
```

Comment:
Not sure why, but the first thing displayed in logs after running this command is `Generating the self-signed certificate
Certificate fingerprint (SHA1): B3:7C:FB:7E:2B:91:12:C2:8B:21:56:25:8B:E2:2A:2D:29:9D:15:53`

Does it mean it didn't use the one provided as volume? (I use the exact same instructions you give for generating self-signed certificate)

Reply:
That's right, is not using the one provided as volume. I'll re-test and check what is wrong.

Reply:
Did you run this command in the same path where the keys were created?

This should print the public key and private key
```bash
cat `pwd`/hawkular-services-public.pem 
cat `pwd`/hawkular-services-private.key
```

Reply:
Looks like I have some issues with mounted volume in the docker container.
My files are there on my local FS but look weird when I check them from inside the container:

```bash
docker exec -it 59d74f014b22 bash
[jboss@59d74f014b22 ~]$ whoami 
jboss
[jboss@59d74f014b22 ~]$ cd /client-secrets/
[jboss@59d74f014b22 client-secrets]$ ls -al
ls: cannot access hawkular-services-public.pem: Permission denied
ls: cannot access hawkular-services-private.key: Permission denied
total 4
drwxr-xr-x.  2 root root   79 Sep 13 09:54 .
drwxr-xr-x. 18 root root 4096 Sep 13 09:54 ..
-??????????  ? ?    ?       ?            ? hawkular-services-private.key
-??????????  ? ?    ?       ?            ? hawkular-services-public.pem
[jboss@59d74f014b22 client-secrets]$ cat hawkular-services-public.pem
cat: hawkular-services-public.pem: Permission denied
[jboss@59d74f014b22 client-secrets]$ 
```

Might be something to do with SELinux

Reply:
yeah problem found: I had to add ":z" flag on mounts (like explained here: https://www.projectatomic.io/blog/2015/06/using-volumes-with-docker-can-cause-problems-with-selinux/ )

Whole command is then:

```bash
docker run --name hawkular-services --link=hawkular-cassandra -e CASSANDRA_NODES=hawkular-cassandra -e HAWKULAR_HOSTNAME=hawkular-services -e HAWKULAR_USE_SSL=true -p 8443:8443 -v `pwd`/hawkular-services-private.key:/client-secrets/hawkular-services-private.key:z -v `pwd`/hawkular-services-public.pem:/client-secrets/hawkular-services-public.pem:z hawkular/hawkular-services
```

Works like a charm after that!